internal/downloader/oci.go

```diff
@@ -0,0 +1,82 @@
+package downloader
+
+import (
+	"github.com/chris-cmsoft/concom/internal"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"os"
+	"path"
+)
+
+type OciDownloader struct {
+	source      string
+	destination string
+
+	// Reference is the processed OCI Name of the Source
+	reference name.Reference
+}
+
+func NewOciDownloader(source, destination string) (Downloader, error) {
+	reference, err := name.ParseReference(source)
+	if err != nil {
+		return nil, err
+	}
+	return &OciDownloader{
+		source:      source,
+		destination: destination,
+		reference:   reference,
+	}, nil
+}
+
+func (dl *OciDownloader) GetFinalDestination() (string, error) {
+	return dl.getOutputDirectory()
+}
+
+// Download executes the download of the OCI artifact into memory, untars it and write it to a directory.
+// This will need to be updated at some point when we are working with OCI artifacts rather than images,
+// to take slightly different actions based on the artifact type we receive from the registry (image / binary / fs)
+func (dl *OciDownloader) Download() error {
+	img, err := remote.Image(dl.reference, remote.WithAuthFromKeychain(authn.DefaultKeychain))
+	if err != nil {
+		return err
+	}
+
+	outputDirectory, err := dl.getOutputDirectory()
+	if err != nil {
+		return err
+	}
+
+	err = os.MkdirAll(outputDirectory, 0755)
+	if err != nil {
+		return err
+	}
+
+	layers, err := img.Layers()
+	for _, layer := range layers {
```

Comment:
consider a wait group if they're massive so you can do each layer concurrently? ie. sync.waitGroup()

Reply:
I considered this during implementation, but purposefully decided to ignore it for the sake of simplicity until it's needed. For now my main focus is readability to make sure it's easy choppable and changeable. WaitGroup is a great improvement, once we know the method is more complete (Still need to figure out OCI artifacts vs. images).

Reply:
Also for the moment with Go plugins, there's just one layer with the binary, and mostly will be just one. This is unlikely to change.